### PR TITLE
Install and configure pgbackrest on patroni cluster

### PR DESCRIPTION
**Pull request by conbrad**
closes: https://github.com/bcgov/wps/issues/2751

- Adjusts patroni Dockerfile and build to include pgbackrest
    - Image build includes object store configuration for pgbackrest to connect and write files
    - Docker image writes pgbackrest global configuration
    - post init script called by patroni enables archive mode / command in postgres, and creates the stanza 

pgbackrest terminology:
- **Stanza**: "A stanza defines the backup configuration for a specific PostgreSQL database cluster."
See: https://pgbackrest.org/configuration.html#section-stanza
- **Archive**: Implies the pushed/pulled WAL segments
See: https://pgbackrest.org/configuration.html#section-archive
- Repo: Implies a collection of one or more stanzas (backup configs for database cluster(s))
See: https://pgbackrest.org/configuration.html#section-repository

# Test Links:
[Landing Page](https://wps-pr-3251.apps.silver.devops.gov.bc.ca/)
[MoreCast 2.0](https://wps-pr-3251.apps.silver.devops.gov.bc.ca/morecast-2)
[Percentile Calculator](https://wps-pr-3251.apps.silver.devops.gov.bc.ca/percentile-calculator)
[MoreCast](https://wps-pr-3251.apps.silver.devops.gov.bc.ca/morecast)
[C-Haines](https://wps-pr-3251.apps.silver.devops.gov.bc.ca/c-haines)
[FireBat](https://wps-pr-3251.apps.silver.devops.gov.bc.ca/fire-behaviour-calculator)
[FireBat bookmark](https://wps-pr-3251.apps.silver.devops.gov.bc.ca/fire-behaviour-calculator?s=266&f=c5&c=NaN&w=20,s=286&f=c7&c=NaN&w=16,s=1055&f=c7&c=NaN&w=NaN,s=305&f=c7&c=NaN&w=NaN,s=344&f=c5&c=NaN&w=NaN,s=346&f=c7&c=NaN&w=NaN,s=328&f=c7&c=NaN&w=NaN,s=1399&f=c7&c=NaN&w=NaN,s=334&f=c7&c=NaN&w=NaN,s=1082&f=c3&c=NaN&w=NaN,s=388&f=c7&c=NaN&w=NaN,s=309&f=c7&c=NaN&w=16,s=306&f=c7&c=NaN&w=NaN,s=1029&f=c7&c=NaN&w=NaN,s=298&f=c7&c=NaN&w=NaN,s=836&f=c7&c=NaN&w=NaN,s=9999&f=c7&c=NaN&w=NaN)
[Auto Spatial Advisory (ASA)](https://wps-pr-3251.apps.silver.devops.gov.bc.ca/auto-spatial-advisory)
[HFI Calculator](https://wps-pr-3251.apps.silver.devops.gov.bc.ca/hfi-calculator)

### CODE REVIEW

File: openshift/patroni-postgis/docker/Dockerfile [hunk 1]
```diff
@@ -3,23 +3,54 @@
...
+    # ca-certificates: for accessing remote raster files;
+    #   fix: https://github.com/postgis/docker-postgis/issues/307
+    ca-certificates \
+    \
+    postgresql-"$PG_MAJOR"-postgis-"$POSTGIS_MAJOR"="$POSTGIS_VERSION" \
+    postgresql-"$PG_MAJOR"-postgis-"$POSTGIS_MAJOR"-scripts \
+    pgbackrest \
     && rm -rf /var/lib/apt/lists/*
 
+RUN mkdir -p -m 777 /var/log/pgbackrest \
```
Comment (conbrad) [inline openshift/patroni-postgis/docker/Dockerfile:24]: Folders for `pgbackrest` are created up front and modified so the non-root user that runs in openshift can access these files when issuing `pgbackrest` commands

File: openshift/patroni-postgis/docker/Dockerfile [hunk 2]
```diff
@@ -3,23 +3,54 @@
...
+    && chmod 777 /etc/pgbackrest/pgbackrest.conf \
+    && chown postgres:postgres /etc/pgbackrest/pgbackrest.conf
+
 # Add the POSTGIS command to the end of the post_init script.
 COPY init_postgis /usr/share/scripts/patroni/
 RUN cat /usr/share/scripts/patroni/init_postgis >> /usr/share/scripts/patroni/post_init.sh
 RUN rm /usr/share/scripts/patroni/init_postgis
 
+RUN echo "[global] \n \
+    repo1-type=s3 \n \
```
Comment (dgboss) [inline openshift/patroni-postgis/docker/Dockerfile:38]: Is the use of `repo1` a pgbackrest thing? Just wondering if a more descriptive name would be useful?

File: openshift/patroni-postgis/docker/init_postgis [hunk 3]
```diff
@@ -1,5 +1,12 @@
 
-if [[ (! -z "$APP_USER") &&  (! -z "$APP_PASSWORD") && (! -z "$APP_DATABASE")]]; then
+if [[ (! -z "$APP_USER") &&  (! -z "$APP_PASSWORD") && (! -z "$APP_DATABASE") && (! -z "$NAME")]]; then
+    echo "Initializing archive mode for pgbackrest"
+    psql -c "alter system set archive_mode=on;"
+    psql -c "alter system set archive_command=pgbackrest --stanza=demo-cluster-1 archive-push '%p';"
```
Comment (dgboss) [inline openshift/patroni-postgis/docker/init_postgis:5]: Wondering about a better name for the stanza?

File: openshift/patroni-postgis/docker/init_postgis [hunk 4]
```diff
@@ -1,5 +1,12 @@
...
     echo "Loading PostGIS extensions into $APP_DATABASE"
     psql "$1" -w -c "\c ${APP_DATABASE}" -c "CREATE EXTENSION IF NOT EXISTS postgis;"
-fi
+    echo "Restarting postgres to init config"
+    pg_ctl restart -D /home/postgres/pgdata/pgroot/data
+    echo "Creating pgbackrest stanza"
+    pgbackrest --stanza="$NAME-cluster-1" --pg1-path=/home/postgres/pgdata/pgroot/data --pg1-port=5432 stanza-create
```
Comment (dgboss) [inline openshift/patroni-postgis/docker/init_postgis:11]: Wondering where/how the `$NAME` gets set? Does `$NAME` have to be `demo` in order to match the stanza parameter on line 5 above?

**Review by dgboss — APPROVED** [review]
Nice work! Thanks for taking this one on!

**Pull request merged**